launch_testing/launch_testing/pytest/plugin.py

```diff
@@ -0,0 +1,528 @@
+
+
+def is_launch_test_mark_valid(item):
+    """
+    Return `True` if the item is a launch test.
+
+    If not, a warning and a skip mark will be added to the item.
+    """
+    kwargs = item.get_closest_marker('launch_testing').kwargs
+    ret = 'fixture' in kwargs and kwargs['fixture'] is not None
+    if not ret:
+        msg = (
+                    '"fixture" keyword argument is required in a pytest.mark.launch_testing() '
+                    f'decorator')
+        item.warn(LaunchTestWarning(msg))
+        item.add_marker(pytest.mark.skip(msg))
+    return ret
+
+
+def has_shutdown_kwarg(item):
+    """Return `True` if the launch test shutdown kwarg is true."""
+    return item.get_closest_marker('launch_testing').kwargs.get('shutdown', False)
+
+
+def get_launch_test_fixture(item):
+    """Return the launch test fixture name, `None` if this isn't a launch test."""
+    mark = item.get_closest_marker('launch_testing')
+    if mark is None:
+        return None
+    try:
+        return mark.kwargs['fixture']
+    except KeyError:
+        return None
+
+
+def get_launch_test_fixturename(item):
+    """Return the launch test fixture name, `None` if this isn't a launch test."""
+    fixture = get_launch_test_fixture(item)
+    return None if fixture is None else fixture.__name__
+
+
+def is_valid_test_item(obj):
+    """Return true if obj is a valid launch test item."""
+    return (
+        inspect.iscoroutinefunction(obj) or inspect.isfunction(obj)
+        or inspect.isgeneratorfunction(obj) or inspect.isasyncgenfunction(obj)
+    )
+
+
+def need_shutdown_test_item(obj):
+    """Return true if we also need to generate a shutdown test item for this object."""
+    return inspect.isgeneratorfunction(obj) or inspect.isasyncgenfunction(obj)
+
+
+def generate_test_items(collector, name, obj, fixturename, is_shutdown, needs_renaming):
+    """Return list of test items for the corresponding object and injects the needed fixtures."""
+    # Inject all needed fixtures.
+    # We use the `usefixtures` pytest mark instead of injecting them in fixturenames
+    # directly, because the second option doesn't handle parameterized launch fixtures
+    # correctly.
+    # We also need to inject the mark before calling _genfunctions(),
+    # so it actually returns many items for parameterized launch fixtures.
+    fixtures_to_inject = (fixturename, 'launch_service', 'event_loop')
+    pytest.mark.usefixtures(*fixtures_to_inject)(obj)
+    items = list(collector._genfunctions(name, obj))
+    for item in items:
+        # Mark shutdown tests correctly
+        item._launch_testing_is_shutdown = is_shutdown
+        if is_shutdown and needs_renaming:
+            # rename the items, to differentiate them from the normal test stage
+            item.name = f'{item.name}[shutdown_test]'
+    return items
+
+
+# Adapted from https://github.com/pytest-dev/pytest-asyncio,
+# see their license https://github.com/pytest-dev/pytest-asyncio/blob/master/LICENSE.
```

Comment:
@ivanpauno meta: not sure how do we go about acknowledging code brought from elsewhere. @tfoote @clalancette ?

Reply:
A few things I would think about:

1. Licensing.  Since the license on pytest-asyncio agrees 100% with this package (Apache 2.0), that works fine.
2. Copyright name.  Typically I would add the Copyright name from the headers in the `pytest-asyncio` repository, but they don't have any.  Since they don't have any attribution there, I don't think we particularly need it.
3. Attribution.  I think attributing it like this is reasonable, though I might point to a specific file at a specific hash rather than just generically at the repository.